File: java/codechallenge/src/main/java/codechallenge/impl/UtilFunc.java
package codechallenge.impl;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class UtilFunc {

	public void formatTimeDiff(final List<Entry> entries, final Date now) {
		entries.forEach(e -> {
			long mills = now.getTime() - e.getTimestamp().getTime();
			long seconds = mills/1000;
			long min = seconds/60;
	        long hours = min/60;
			
	        String res = new String();
	        
	        if(hours > 0)  {
	        	if(hours == 1) res = String.format("%d hour ago", hours);
	        	else res = String.format("%d hours ago", hours);
	        } else if( min > 0) {
	        	if(min == 1) res = String.format("%d minute ago", min);
	        	else res = String.format("%d minutes ago", min);
	        } else {
	        	if(seconds == 1) res = String.format("%d second ago", seconds);
	        	else res = String.format("%d seconds ago", seconds);
	        }

	        e.setTimediff(res);
		});
	}
	
	public List<String> formatEntries(final List<Entry> entries) {
		return formatEntriesSpec(entries, false);
	}
	
	public List<String> formatEntriesWithUser(final List<Entry> entries) {
		return formatEntriesSpec(entries, true);
	}
	
	private List<String> formatEntriesSpec(final List<Entry> entries, boolean withName) {
		List<String> res = new ArrayList<String>(entries.size());
		
		entries.forEach(e -> {
			if(withName) res.add( String.format("%s - %s (%s)", e.getUser(), e.getMessage(), e.getTimediff()) );
			else res.add( String.format("%s (%s)", e.getMessage(), e.getTimediff()) );
		});
		
		return res;
	}
}

File: cpp/logic/src/utilfunc.cpp
#include <sstream>

#include "entry.h"
#include "utilfunc.h"

std::vector<std::string> formatEntries(const std::vector<Entry>& entries, bool withName) {
    std::vector<std::string> ret;

    for( Entry e : entries ) {
        std::stringstream res;

        if(withName) res << e.user << " - ";

        res << e.message << " (" << e.timediff << ")";

        ret.push_back(res.str());
    }

    return ret;
}

void formatTimeDiff(std::vector<Entry>& entries) {
    for( Entry& e : entries ) {

        int seconds = std::difftime(std::time(nullptr), e.timestamp);

        int min = seconds/60;
        int hours = min/60;

        std::stringstream res;

        if(hours > 0) hours == 1 ? res << hours << " hour ago" : res << hours << " hours ago";
        else if( min > 0) min == 1 ? res << min << " minute ago" : res << min << " minutes ago";
        else seconds == 1 ? res << seconds << " second ago" : res << seconds << " seconds ago";

        e.timediff = res.str();
    }
}
